### ISSUE TITLE
Module compilation leaks memory?

### ISSUE THREAD

**Issue by Rochet2** [WAVM/WAVM#293]
My program's memory consumption is increasing when loading and freeing modules despite trying to free all structures related to a module.
I have tracked the source to the compilation step, which seems to spend 5MB of RAM for my module and never free it until the entire program exits.

I made a few tests with the embedder-example in `Examples/embedder/cpp-wasi/embedder-example.cpp` that replicate the issue.

Using e289ba654fe51655ab080819fd85ed9c936aae6e with and without the following patch: https://gist.github.com/Rochet2/a55cdded21d172a04a758caa79e63e7a

| Execution description                                                                                                                                           | RAM    |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------|--------|
| Initially at line https://github.com/WAVM/WAVM/blob/master/Examples/embedder/cpp-wasi/embedder-example.cpp#L80                                              | 5MB    |
| At line https://github.com/WAVM/WAVM/blob/master/Examples/embedder/cpp-wasi/embedder-example.cpp#L92                                                        | 5MB    |
| At line https://github.com/WAVM/WAVM/blob/master/Examples/embedder/cpp-wasi/embedder-example.cpp#L100                                                       | 12.9MB |
| When instantiating the module and running the code until line https://github.com/WAVM/WAVM/blob/master/Examples/embedder/cpp-wasi/embedder-example.cpp#L157 | 13MB   |
| At line https://github.com/WAVM/WAVM/blob/master/Examples/embedder/cpp-wasi/embedder-example.cpp#L161                                                       | 12.9MB |

From the above table, it seems that the compilation of a module takes memory and never releases it.
The GC step at the end probably works and seems to release the instantiated parts.
The addition of the brackets in the patch/diff above will ensure that all allocated structures within the stack are freed before we measure the memory at program exit. However, memory consumption stays the same regardless of them.

I also tried running valgrind and here are the results: https://gist.github.com/Rochet2/87666d32062014816d0020eb58e0b9d9
According to it some parts of the compilation could possibly depend on uninitialized values. The valgrind log is with the patch applied.

Here is the module I tested with: [main.zip](https://github.com/WAVM/WAVM/files/5561060/main.zip)

Is this a bug or is it expected for compilation to use memory indefinitely?

**Comment by AndrewScheidecker** [comment]
> According to it some parts of the compilation could possibly depend on uninitialized values.

I see the issue. The code is technically correct, but I'll change it to avoid this warning.

> Is this a bug or is it expected for compilation to use memory indefinitely?

I'm not ruling out that there is a bug here, but it's expected for a process not to free memory to the OS immediately after the memory is freed for various reasons: among them fragmentation, and optimizing for the case that the memory will soon be allocated again. I assume that's what is happening here.

If you set `ALLOCATE_SIGALTSTACK_ON_MAIN_STACK` to 0 [here](https://github.com/WAVM/WAVM/blob/7d865f618cbbb916240b116c64a335ce6ff43abc/Lib/Platform/POSIX/ThreadPOSIX.cpp#L56), the program should run under valgrind without a stack overflow signal. If I run the cpp-wasi embedder example with that change under valgrind, I see this:
```
==4386== LEAK SUMMARY:
==4386==    definitely lost: 0 bytes in 0 blocks
==4386==    indirectly lost: 0 bytes in 0 blocks
==4386==      possibly lost: 0 bytes in 0 blocks
==4386==    still reachable: 138,713 bytes in 1,038 blocks
==4386==         suppressed: 0 bytes in 0 blocks
```

**Comment by Rochet2** [comment]
Thanks. Closing this as new tests on 12c06f113e99a11af3b39fc8181056c1a55c27b6 show that there were no issues or they were fixed.
Also, you fixed the issues Valgrind was complaining about already.

I tested running a program that in a while loop continuously compiles the same module and then frees any related variables. Doing so for 15 minutes maintains the same memory footprint according to `top` command and the ubuntu system monitor.
Additionally, I ran Valgrind on a program that compiled and then freed N modules. Increasing N by an order of magnitude each time had no effect on the `still reachable` value in valgrind and all other values were 0.
